Allow to easily bisect flaky failures

deivid-rodriguez opened the pull request:
# Description:

I used this today to bisect some flaky test failures. I thought I should share it in case other rubygems devs find it useful.

For example,

```
SEED=26796 rake
```

currently fails with the error described in #2625.

Running

```
SEED=26796 rake bisect
```

should give you the minimal reproduction for it after a bit of investigation, namely,

```
SEED=26796 rake TESTOPTS="--name=/test_generate_index_modern_back_to_back\|test_doctor_dry_run/"
```

# Tasks:

- [ ] Describe the problem / feature
- [ ] Write tests
- [ ] Write code to solve the problem
- [ ] Get code review from coworkers / friends

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

simi commented on util/bisect:
```diff
@@ -0,0 +1,20 @@
+
+require 'path_expander'
+
+class PathExpander
```
Is this the only change needed to successfully run `minitest_bisect`?

deivid-rodriguez replied:
`minitest_bisect` runs fine without this. The problem is that `rake` and `path_expander` (`minitest-bisect`'s file loader) currently load files differently. So, they shuffle a different array when randomizing. Thus, failures reproducible through `rake` don't reproduce through `minitest_bisect` even if the same seed is used.

zenspider replied:
Are you on Linux or something such that `Dir["*"]` isn't sorted? Is rake explicitly sorting?

deivid-rodriguez replied:
Yes, I'm on Linux. Rake explicitly sorts [here](https://github.com/ruby/rake/blob/67ced2fab148db0ada6b0d52d321f5cc86e22960/lib/rake/file_list.rb#L404-L409), so for example in rubygems I get

```
> irb(main):001:0> FileList['test/**/test_*.rb']
=> ["test/rubygems/test_bundled_ca.rb", "test/rubygems/test_config.rb", "test/rubygems/test_deprecate.rb", "test/rubygems/test_gem.rb", "test/rubygems/test_gem_available_set.rb", ..., "test/rubygems/test_gem_version.rb", "test/rubygems/test_gem_version_option.rb", "test/rubygems/test_kernel.rb", "test/rubygems/test_remote_fetch_error.rb", "test/rubygems/test_require.rb"]

irb(main):002:0> Dir['test/**/test_*.rb']
=> ["test/rubygems/test_gem_commands_info_command.rb", "test/rubygems/test_gem_resolver_vendor_set.rb", "test/rubygems/test_gem_dependency_list.rb", "test/rubygems/test_gem_source_fetch_problem.rb", "test/rubygems/test_gem_request.rb", "test/rubygems/test_gem_resolver_best_set.rb", ..., "test/rubygems/test_gem_gem_runner.rb", "test/rubygems/test_gem_resolver_local_specification.rb", "test/rubygems/test_gem_available_set.rb", "test/rubygems/test_gem_security_policy.rb", "test/rubygems/test_gem_security.rb", "test/rubygems/test_gem_impossible_dependencies_error.rb"]
```

simi commented:
I don't think this should be part of rubygems itself. There will be problem maintaining this IMHO (for example there's already hardcoded path to minitest server 1.0.5).

If there's any problem running `minitest-bisect` in this codebase without monkey patching path `path_expander`, isn't that actually problem in `minitest-bisect` gem (or `path_expander` gem)?

deivid-rodriguez commented:
Yes! I needed both changes to get this working, but maybe there are better ways. Can you try this in your environment?

The `path_expander` changes can definitely be incorporated into the `path_expander` gem. I just wanted to try get this working without having to (yet) open PRs to external gems. Maybe I can politely ping @zenspider and ask for his opinion?

The hardcoded route to `minitest-server` is harder to avoid I think. Minitest plugin loading process looks for files with specific names (`minitest/<name_of_plugin>_plugin.rb`) in the `$LOAD_PATH`, and `minitest-server` is implemented as a `minitest` plugin. So... not sure how to do this without using `bundler`... Maybe we should just use `bundler` :sweat_smile:.

djberg96 commented on Rakefile:
```diff
@@ -71,6 +71,18 @@ task(:rubocop) do
+desc "Run a test suite bisection"
+task(:bisect) do
+  seed = begin
+           Integer(ENV["SEED"])
```
I'd vote to let the task take an argument instead of relying on an ENV setting. Or you could do both:

```
task(:bisect, [:seed]) do
   seed ||= ENV['SEED']
   # ...
end
```
Then call it like `rake bisect[12345]`.

deivid-rodriguez replied:
That's neat, I'll add that!

deivid-rodriguez replied:
I was about to add this but I'm now not really sure I prefer it. Given that for running tests we do `SEED=12345 rake test`, but we can't do `rake test[12345]`. Doesn't it make more sense to keep it consistent?

djberg96 replied:
Well, as I said, you can support either approach by defaulting to ENV['SEED'] if no arg is provided, so CI would still work. But, it's your PR. :)

deivid-rodriguez commented:
How do other devs feel about this? I think it's very useful, I used it again in #2638.

deivid-rodriguez commented:
By the way, I created a PR to PathExpander, hopefully the monkeypatch can be removed in the future!

hsbt commented:
It seems useful for me too.

hsbt commented:
@bundlerbot r+

deivid-rodriguez commented:
@hsbt I got caught by the cops :smile:

I rebased the PR and fixed the style issues, and also added a comment about the `path_expander` monkeypatching and improved the error message when `minitest-bisect` is not installed.

I think it should be ready now!

hsbt commented:
Thanks a lot!

hsbt commented:
@bundlerbot retry

deivid-rodriguez commented:
Thanks to you! :heart_eyes:
